### PR TITLE
CQL2 - implements the Abstract TestSuite (ATS) conformance testsuite

### PR DESCRIPTION
**Pull request by pmauduit**
See https://github.com/geoserver/geoserver-ogcapi/issues/10 for the motivation. This PR mainly provides a transcription of the OGC Abstract Test Suite to Java test classes for both cql2 modules (text & json).

It also reflects some changes from the specifications, e.g. renaming ENVELOPE to BBOX in the filters. Thanks to @aaime, it also:
* introduces a 3 way logic evalution via datastore hints
* fixes a parsing issue in the envelopes
* adds support for extra arithmetic operators, %, ^ and div

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [x] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [ ] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

File: modules/unsupported/cql2-json/src/test/java/org/geotools/filter/text/cqljson/conformance/ConformanceTest17OnlineTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+/**
+ * See <a href="https://docs.ogc.org/is/21-065r2/21-065r2.html#_conformance_test_17">table 10 from
+ * section A.5.3 Conformance test 17.</a>
+ */
+@RunWith(Parameterized.class)
+@Ignore("Not implemented yet")
+public class ConformanceTest17OnlineTest extends ATSOnlineTest {
```
Comment (aaime) [inline modules/unsupported/cql2-json/src/test/java/org/geotools/filter/text/cqljson/conformance/ConformanceTest17OnlineTest.java:39]: Shouldn't it extend ConformanceTest17OnlineTest ?

File: modules/unsupported/cql2-json/src/test/java/org/geotools/filter/text/cqljson/conformance/ConformanceTest17OnlineTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        this.criteria = criteria;
+    }
+
+    @Override
+    protected Filter criteriaToFilter(String criteria) throws CQLException {
+        return CQL2Json.toFilter(criteria);
+    }
+
+    @Parameterized.Parameters(name = "{index} {0}")
+    public static Collection<Object[]> params() {
```
[the remark this answers is not in the document]
Comment (aaime) [reply modules/unsupported/cql2-json/src/test/java/org/geotools/filter/text/cqljson/conformance/ConformanceTest17OnlineTest.java:54]: Eventually the explanations could be centralized in one place... dunno, the ConformanceUtils class for example. And just point there any time filter references are missing.

File: modules/unsupported/cql2-json/src/test/java/org/geotools/filter/text/cqljson/conformance/ConformanceTest23OnlineTest.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    @Override
+    protected Filter criteriaToFilter(String criteria) throws CQLException {
+        return CQL2Json.toFilter(criteria);
+    }
+
+    @Parameterized.Parameters(name = "{index} {0}")
+    public static Collection<Object[]> params() {
+        return Arrays.asList(
```
Comment (aaime) [inline modules/unsupported/cql2-json/src/test/java/org/geotools/filter/text/cqljson/conformance/ConformanceTest23OnlineTest.java:55]: See ConformanceTest17OnlineTest comments

File: modules/unsupported/cql2-json/src/test/java/org/geotools/filter/text/cqljson/conformance/ConformanceTest42OnlineTest.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                    {
+                        "{\"op\":\"t_before\",\"args\":[{\"property\":\"date\"},{\"date\":\"2022-04-16T00:00:00GMT+00:00\"}]}",
+                        1
+                    },
+                    // {"{}", 2}, // TODO broken text filter, cannot be converted to json
+                    {
+                        "{\"op\":\"t_equals\",\"args\":[{\"property\":\"date\"},{\"date\":\"2022-04-16T00:00:00GMT+00:00\"}]}",
+                        1
+                    },
+                    /* {"{}", 1}, // TODO broken text filter, cannot be converted to json
```
Comment (aaime) [inline modules/unsupported/cql2-json/src/test/java/org/geotools/filter/text/cqljson/conformance/ConformanceTest42OnlineTest.java:66]: See ConformanceTest17OnlineTest comments

File: pom.xml [hunk 5]
```diff
@@ -1811,6 +1811,9 @@
           <redirectTestOutputToFile>false</redirectTestOutputToFile>
           <forkMode>${test.forkMode}</forkMode>
           <trimStackTrace>false</trimStackTrace>
+          <environmentVariables>
```
Comment (aaime) [inline pom.xml:1814]: I would add a clarifying comment as to why the env variable is set

**Review by aaime — CHANGES_REQUESTED** [review]
Great job @pmauduit, I left a few comments below.

**Comment by pmauduit** [comment]
thanks for the feedback, PR updated consequently

**Pull request merged**